app/src/main/kotlin/com/hippo/ehviewer/ktor/Apache5Extensions.kt

```diff
@@ -28,7 +41,42 @@ fun Apache5EngineConfig.configureClient() {
                         setIoThreadCount(1)
                     }.build(),
                 )
+                if (isAtLeastQ) {
+                    val resolver = DnsResolver.getInstance()
+                    val executor = Dispatchers.IO.asExecutor()
+                    setDnsResolver(
+                        object : ApacheDnsResolver {
+                            override fun resolve(host: String) = query(host, resolver, executor).toTypedArray()
+                            override fun resolveCanonicalHostname(host: String) = unreachable()
+                        },
+                    )
+                }
             }.build(),
         )
     }
 }
+
+@RequiresApi(Build.VERSION_CODES.Q)
+fun query(host: String, resolver: DnsResolver, executor: Executor): List<InetAddress> = runBlocking {
+    try {
+        suspendCoroutine { cont ->
+            resolver.query(
+                null,
+                host,
+                DnsResolver.FLAG_EMPTY,
+                executor,
+                null,
+                object : DnsResolver.Callback<List<InetAddress>> {
+                    override fun onAnswer(result: List<InetAddress>, r: Int) {
+                        cont.resume(result)
+                    }
+                    override fun onError(e: DnsResolver.DnsException) {
+                        cont.resumeWithException(e)
+                    }
+                },
+            )
```

Comment:
For some reason the connectivity check will fail when used with VPN.

Reply:
Actually I don't believe this will improve performance as we use something non-blocking as blocking for blocking usage.
But when I measure throughput & latency performance, both apache5 & cronet give a not ideal result.
The only thing we know is 1 IO thread is not the bottleneck as it takes 80% time to poll but not handle input/output from flame graph.

Reply:
So I decided to suspend this MR